### PR TITLE
Upgrade to microgrid client v0.4.0

### PR DESCRIPTION
**Pull request by llucax**
- **Bump frequenz-client-microgrid to v0.4.0**
- **Use `GrpcError` instead of `grpc.aio.AioRpcError`**
- **Avoid repetition on error handling**
- **Use a server URL to connect to the microgrid API**
- **Use new microgrid API wrapped types**
- **Stop using the `mock_api` for testing**
- **Remove the, now unused, `mock_api` and its tests**
- **Silence `hpack` debug logging**
- **Fix mock return value**
- **Avoid disabling some pylint checks when unnecessary**
- **Remove unsed protobuf and grpcio dependencies**

### CODE REVIEW

**Comment by llucax** [comment]
This is a draft because it depends on https://github.com/frequenz-floss/frequenz-client-microgrid-python/pull/54 and v0.4.0 to be released after it is merged.

~It also needs release notes~, but other than that it is ready to be reviewed!

**Comment by llucax** [comment]
Added release notes.

**Comment by llucax** [comment]
v0.4.0 released, this is ready for a final review!

File: tests/microgrid/test_microgrid_api.py
Comment (daniel-zullo-frequenz) [inline tests/microgrid/test_microgrid_api.py:107]: this probably goes in another commit (different than `Use a server URL to connect to the microgrid API`

File: tests/actor/test_data_sourcing.py
Comment (daniel-zullo-frequenz) [inline tests/actor/test_data_sourcing.py:148]: I think it is not a good idea to use the `==` operator to compare a floating-point value against the result of an addition operation involving floating-point numbers, even in tests. I'd suggest to use isclose() from pytest or math instead

File: tests/actor/test_data_sourcing.py
Comment (daniel-zullo-frequenz) [inline tests/actor/test_data_sourcing.py]: looks like `Use new microgrid API wrapped types` is not the right commit for this change

File: pyproject.toml
Comment (daniel-zullo-frequenz) [inline pyproject.toml:192]: The changes in this file  seem not related to `Avoid disabling some pylint checks when unnecessary`

File: mkdocs.yml
Comment (daniel-zullo-frequenz) [inline mkdocs.yml:123]: hmm this also doesn't seem to be related to the commit `Avoid disabling some pylint checks when unnecessary`

File: RELEASE_NOTES.md
Comment (daniel-zullo-frequenz) [inline RELEASE_NOTES.md]: I'd avoid the double `too` in the last sentence
```suggestion
- The `frequenz-client-microgrid` dependency was bumped to v0.4.0. If you are using the client directly in your code, you will need to upgrade too.
```

**Review by daniel-zullo-frequenz — COMMENTED** [review]
I have a few minor suggestions to check for, LGTM otherwise

**Review by daniel-zullo-frequenz — COMMENTED** [review]
I have a few minor suggestions to check for, LGTM otherwise

File: tests/microgrid/test_microgrid_api.py
Comment (llucax) [reply tests/microgrid/test_microgrid_api.py:107]: It is related, because the parsing of the URL will now return a `grpc.client.Channel` and before doing that switch, we were using a `grpc.aio.insecure_channel`. So we need to adapt the tests to work with this change.

File: tests/actor/test_data_sourcing.py
Comment (llucax) [reply tests/actor/test_data_sourcing.py:148]: Can you please create a separate issue? I'm just changing the indentation to this code.

File: tests/actor/test_data_sourcing.py
Comment (llucax) [reply tests/actor/test_data_sourcing.py]: Moved to the commit introducing the `server_url` :+1:

File: pyproject.toml
Comment (llucax) [reply pyproject.toml:192]: Moved to a new commit :+1:

File: mkdocs.yml
Comment (llucax) [reply mkdocs.yml:123]: Moved to a new commit :+1:

File: RELEASE_NOTES.md
Comment (llucax) [reply RELEASE_NOTES.md]: Updated, thanks!

**Comment by llucax** [comment]
Updated.

File: src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py
Comment (shsms) [inline src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py:670]: off topic, but just noticing how these are implemented for battery power distribution.  The logs suggest that for "OutOfRange" errors, we don't mark anything as broken, but for other errors, we mark them as broken, but I'm not sure if that's the case.  I think they're all marked as broken.  Need to confirm.

Also, when battery is over SoC, and we send a power request, the request would be out of bounds, but still the api service sends "FAILED_PRECONDITION", which doesn't sound like something we can ignore.  If it would send OutOfRange instead, we can avoid blocking batteries unnecessarily just because they're fully charged.

File: src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py
Comment (shsms) [inline src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py]: "error" is not a verb, treating it like it is feels really weird.  Other options: `oper_success = False`, `has_error` = True, `failed = True`, `succeeded = False`.

I like `succeeded = False`, `op_successful`, etc., because they seem to presume less.

File: examples/battery_pool.py
Comment (shsms) [inline examples/battery_pool.py:26]: could maybe be configurable through the client?

File: pyproject.toml
Comment (shsms) [inline pyproject.toml]: `tqdm` and `watchfiles` are unused dependencies as well, maybe they can go away too, along with these.

File: src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py:670]: I think ignoring a failed precondition could be fine, it just means there something else that needs to be done before retrying could succeed:

> The operation was rejected because the system is not in a state required for the operation's execution. For example, the directory to be deleted is non-empty, an rmdir operation is applied to a non-directory, etc. Service implementors can use the following guidelines to decide between FAILED_PRECONDITION, ABORTED, and UNAVAILABLE: (a) Use UNAVAILABLE if the client can retry just the failing call. (b) Use ABORTED if the client should retry at a higher level (e.g., when a client-specified test-and-set fails, indicating the client should restart a read-modify-write sequence). (c) Use FAILED_PRECONDITION if the client should not retry until the system state has been explicitly fixed. E.g., if an "rmdir" fails because the directory is non-empty, FAILED_PRECONDITION should be returned since the client should not retry unless the files are deleted from the directory. 

Out of range is considered some sort of special case of failed precondition:

> There is a fair bit of overlap between FAILED_PRECONDITION and OUT_OF_RANGE. We recommend using OUT_OF_RANGE (the more specific error) when it applies so that callers who are iterating through a space can easily look for an OUT_OF_RANGE error to detect when they are done. 

I agree out of range could make more sense if the request fails because the power is out of bounds.

File: src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py]: Really? Then it is Internet English, I have seen it used many times. Also Wiktionary has an entry for it: https://en.wiktionary.org/wiki/errored, and some sites even has the conjugation for it: https://www.verbix.com/webverbix/english/error. The Oxford dictionary seems to have an entry for it too: https://www.oed.com/dictionary/error_v?tab=factsheet#5217607, it seems to be used since 1820 :upside_down_face: (but I couldn't find it in the Cambridge dictionary).

I can change it anyways, I was only surprised that it might not be as common as I thought.

File: examples/battery_pool.py
Comment (llucax) [reply examples/battery_pool.py:26]: Mmm, though one. I really don't think we should mess with silencing logging levels or loggers in libraries, IMHO that should be up to the end user.

Comment (llucax) [reply examples/battery_pool.py:26]: Also, please note we are enabling `DEBUG` logging in the above line, I'm actually surprised we are not getting much more crap from that setting. Normally you would have to do a much more conscious configuring of the logging system. I think the SDK eventually could provide some sensible default logging config (and probably a way to update the logging config dynamically, via `ConfigManager`), and I would be fine to set `hpack` logging to `INFO` there by default.

File: src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py]: Changed to `failed = True` because then we are setting `failed_xxx` variables, so it looks pretty consistent.

File: pyproject.toml
Comment (llucax) [reply pyproject.toml]: Actually `timezonefinder` is not used anymore either (it was moved to the microgrid API). Added a new commit removing dependencies not used for a long time now.

**Comment by llucax** [comment]
Updated.

File: src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py
Comment (shsms) [reply src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py]: [Chambers](https://chambers.co.uk/search/?query=error&title=21st) and [Webster](https://www.merriam-webster.com/dictionary/error) also list only the noun form.  OED must have been smoking something.

But `succeeded = False` might be better, because that doesn't imply there was an error, and in context, it just means we don't know yet, unlike `error = True`.

**Review by shsms — APPROVED** [review]

File: src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py
Comment (llucax) [reply src/frequenz/sdk/actor/power_distributing/_component_managers/_battery_manager.py]: I was going to change it but I see you already approved, so I will keep `failed` to avoid the extra review roundtrip as it is such a small detail, I guess it is fine to keep it with `failed`.

File: tests/microgrid/test_microgrid_api.py
Comment (daniel-zullo-frequenz) [reply tests/microgrid/test_microgrid_api.py:107]: Oh I see, sorry for the noise then

File: tests/actor/test_data_sourcing.py
Comment (daniel-zullo-frequenz) [reply tests/actor/test_data_sourcing.py:148]: ok, will create an issue

Comment (daniel-zullo-frequenz) [reply tests/actor/test_data_sourcing.py:148]: https://github.com/frequenz-floss/frequenz-sdk-python/issues/965

**Pull request merged**
